File: algorithms/leet.1734.src.1.py
class Solution:
    def decode(self, encoded: List[int]) -> List[int]:
        n = len(encoded) + 1
        encoded.append(reduce(lambda a, b: a^b, encoded))
        neg = []
        # print(encoded)
        x = reduce(lambda a, b: a^b, encoded[1::2])
        y = reduce(lambda a, b: a^b, encoded[2::2])
        # print(x, y)
        for i in range(n):
            neg.append(x)
            # print(encoded[(i+1)%n], encoded[(n+i)%n])
            x ^= encoded[(i+1)%n]
            x ^= encoded[(n+i)%n]
            x, y = y, x
        k = reduce(lambda a, b: a^b, range(1, n+1))
        return [k ^ x for x in neg]

File: algorithms/leet.1734.src.1.cpp
class Solution {
public:
    vector<int> decode(vector<int>& encoded) {
        int n = encoded.size() + 1;
        int x = 0;
        int y = 0;
        for (int i = 0; i < n-1; ++i) {
            x ^= encoded[i];
        }
        encoded.push_back(x);
        vector<int> ans(0);
        // k 是全集异或
        int k = 0;
        for (int i = 0; i < n; ++i) {
            k ^= i + 1;
        }
        x = y = 0;
        for (int i = 1; i < n; i += 2) {
            x ^= encoded[i];
            y ^= encoded[i+1];
        }
        for (int i = 0; i < n; ++i) {
            ans.push_back(x ^ k);
            x ^= encoded[(i+1)%n];
            x ^= encoded[(i+n)%n];
            swap(x, y);
        }
        return ans;
    }
};
